src/models/tools/table/table-content.ts

```diff
@@ -204,101 +235,121 @@ export const TableContentModel = ToolContentModel
-      self.linkedGeometries.clear();
-    },
+      // if (self.metadata.hasExpressions) {
+      //   self.metadata.updateDatasetByExpressions(self.dataSet);
+      // }
+    }
   }))
   .actions(self => ({
+    clearImportedDataSet() {
+      self.importedDataSet = DataSet.create();
+    },
     logChange(change: ITableChange) {
       const toolId = self.metadata?.id || "";
       Logger.logToolChange(LogEventName.TABLE_TOOL_CHANGE, change.action, change, toolId);
     }
   }))
   .actions(self => ({
+    afterAttach() {
+      // Monitor our parents and update our shared model when we have a document parent
+      addDisposer(self, reaction(() => {
+        const sharedModelManager = self.tileEnv?.sharedModelManager;
+
+        const sharedDataSet = sharedModelManager?.isReady
+          ? sharedModelManager?.findFirstSharedModelByType(SharedDataSet, self.metadata.id)
+          : undefined;
+
+        const tileSharedModels = sharedModelManager?.isReady
+          ? sharedModelManager?.getTileSharedModels(self)
+          : undefined;
+
+        return { sharedModelManager, sharedDataSet, tileSharedModels };
+      },
+      // reaction/effect
+      ({sharedModelManager, sharedDataSet, tileSharedModels}) => {
+        if (!sharedModelManager?.isReady) {
+          // We aren't added to a document yet so we can't do anything yet
+          return;
+        }
+
+        if (sharedDataSet && tileSharedModels?.includes(sharedDataSet)) {
+          // The shared model has already been registered by a client, but as the
+          // "owner" of the data, we synchronize it with our local content.
+          if (!self.importedDataSet.isEmpty) {
+            sharedDataSet.dataSet = cloneDeep(self.importedDataSet) as unknown as IDataSet;
+            self.clearImportedDataSet();
+          }
+        }
+        else {
+          if (!sharedDataSet) {
+            // The document doesn't have a shared model yet
+            const dataSet = !self.importedDataSet.isEmpty ? cloneDeep(self.importedDataSet) : DataSet.create();
+            self.clearImportedDataSet();
+            sharedDataSet = SharedDataSet.create({ providerId: self.metadata.id, dataSet });
+          }
+
+          // Add the shared model to both the document and the tile
+          sharedModelManager.addTileSharedModel(self, sharedDataSet);
+        }
+      },
+      {name: "sharedModelSetup", fireImmediately: true}));
+    },
+    updateAfterSharedModelChanges(sharedModel?: SharedModelType) {
+      console.warn("updateAfterSharedModelChanges hasn't been implemented for table content.");
+    },
     setTableName(name: string) {
-      self.dataSet.name = name;
+      self.dataSet.setName(name);
 
-      self.logChange({
-              action: "update",
-              target: "table",
-              props: { name }
-            });
+      self.logChange({ action: "update", target: "table", props: { name } });
```

Comment:
This change could be logged automatically along with all other actions on the table-content by using `onTileAction`. You can see it in use in `drawing-content.ts`. Probably that change should go in another PR though.